Tests/AsyncHTTPClientTests/AsyncTestHelpers.swift

```diff
@@ -0,0 +1,189 @@
+class AsyncSequenceWriter: AsyncSequence {
+    typealias AsyncIterator = Iterator
+    typealias Element = ByteBuffer
+
+    struct Iterator: AsyncIteratorProtocol {
+        typealias Element = ByteBuffer
+
+        private let writer: AsyncSequenceWriter
+
+        init(_ writer: AsyncSequenceWriter) {
+            self.writer = writer
+        }
+
+        mutating func next() async throws -> ByteBuffer? {
+            try await self.writer.next()
+        }
+    }
+
+    func makeAsyncIterator() -> Iterator {
+        return Iterator(self)
+    }
+
+    private enum State {
+        case buffering(CircularBuffer<ByteBuffer?>, CheckedContinuation<Void, Error>?)
+        case finished
+        case waiting(CheckedContinuation<ByteBuffer?, Error>)
+        case failed(Error)
+    }
+
+    private var _state = State.buffering(.init(), nil)
+    private let lock = Lock()
+
+    public var hasDemand: Bool {
+        self.lock.withLock {
+            switch self._state {
+            case .failed, .finished, .buffering:
+                return false
+            case .waiting:
+                return true
+            }
+        }
+    }
+
+    public func demand() async throws {
+        self.lock.lock()
+
+        switch self._state {
+        case .buffering(let buffer, .none):
+            try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
+                self._state = .buffering(buffer, continuation)
+                self.lock.unlock()
+            }
+
+        case .waiting:
+            self.lock.unlock()
+            return
+
+        case .buffering(_, .some):
+            self.lock.unlock()
+            preconditionFailure("Already waiting for demand")
+
+        case .finished, .failed:
+            let state = self._state
+            self.lock.unlock()
+            preconditionFailure("Invalid state: \(state)")
+        }
+    }
+
+    private func next() async throws -> ByteBuffer? {
+        self.lock.lock()
+        switch self._state {
+        case .buffering(let buffer, let demandContinuation) where buffer.isEmpty:
+            return try await withCheckedThrowingContinuation { continuation in
+                self._state = .waiting(continuation)
+                self.lock.unlock()
+                demandContinuation?.resume(returning: ())
+            }
+
+        case .buffering(var buffer, let demandContinuation):
+            let first = buffer.popFirst()!
+            if first != nil {
+                self._state = .buffering(buffer, demandContinuation)
+            } else {
+                self._state = .finished
+            }
+            self.lock.unlock()
+            return first
+
+        case .failed(let error):
+            self._state = .finished
+            self.lock.unlock()
+            throw error
+
+        case .finished:
+            return nil
+
+        case .waiting:
+            preconditionFailure("How can this be called twice?!")
```

Comment:
I generally prefer precondition failure message to be clearer about what actually failed but not a big deal for test utils